Sources/SBTUITestTunnelServer/private/SBTProxyURLProtocol.m

```diff
@@ -807,4 +807,39 @@ - (NSDictionary *)blockCookieRuleFromMatchingRules:(NSArray<NSDictionary *> *)ma
     return nil;
 }
 
+// NSURLProtocol emits a runtime warning when a non-plist type is given to `setProperty`:
+//
+//    API MISUSE: properties set by +[NSURLProtocol setProperty:forKey:inRequest:] should only include property
+//    list types (NSArray, NSDictionary, NSString, NSData, NSDate, NSNumber).
+//
+// The methods below perform serialization for NSURLRequest types before/after they are sent through NSURLProtocol.
+
+/// Finds the original request in NSURLProtocol and deserializes it
++ (NSURLRequest *)originalRequestFor:(NSURLRequest*)request {
+    NSData *serializedOriginal = [NSURLProtocol propertyForKey:SBTProxyURLOriginalRequestKey inRequest:request];
+    NSURLRequest *originalRequest = nil;
+
+    if (serializedOriginal) {
+        // needs to be deserialized after retrieving from NSURLProtocol
+        NSError *unarchiveError;
+        NSSet *classes = [NSSet setWithObjects:[NSURLRequest class], [NSMutableURLRequest class], nil];
+        originalRequest = [NSKeyedUnarchiver unarchivedObjectOfClasses:classes fromData:serializedOriginal error:&unarchiveError];
+        NSAssert(unarchiveError == nil, @"Error unarchiving NSURLRequest from NSURLProtocol");
+    }
+
+    return originalRequest;
+}
+
+/// Associates the original request to the current request by serializing and storing it
++ (void)associateOriginalRequest:(NSURLRequest *)original withRequest:(NSMutableURLRequest*)request {
```

Comment:
Both of these are class methods because one use was from inside `+matchingRulesForRequest:`